src/spesh/disp.c

```diff
@@ -1057,6 +1110,217 @@ static int translate_dispatch_program(MVMThreadContext *tc, MVMSpeshGraph *g,
+                            case MVM_NATIVECALL_ARG_VOID:
+                                base_op = MVM_op_get_op(MVM_OP_sp_runnativecall_v);
+                                keep_result_register = 0;
+                                has_return_value = 0;
+                                break;
+                            case MVM_NATIVECALL_ARG_CHAR:
+                            case MVM_NATIVECALL_ARG_SHORT:
+                            case MVM_NATIVECALL_ARG_INT:
+                            case MVM_NATIVECALL_ARG_LONG:
+                            case MVM_NATIVECALL_ARG_LONGLONG:
+                            case MVM_NATIVECALL_ARG_CPOINTER:
+                                base_op = MVM_op_get_op(MVM_OP_sp_runnativecall_i);
+                                box_return_value = 1;
+                                break;
+                        }
+                    }
+                }
+
+                /* Form the varargs op and create the instruction. */
+                MVMOpInfo *rb_op = MVM_spesh_alloc(tc, g, MVM_spesh_disp_dispatch_op_info_size(
+                    tc, base_op, callsite));
+                MVM_spesh_disp_initialize_dispatch_op_info(tc, base_op, callsite, rb_op);
+                MVMSpeshIns *rb_ins = MVM_spesh_alloc(tc, g, sizeof(MVMSpeshIns));
+                rb_ins->info = rb_op;
+                rb_ins->operands = MVM_spesh_alloc(tc, g,
+                     rb_op->num_operands * sizeof(MVMSpeshOperand));
+
+                /* Write result into dispatch result register unless void. */
+                MVMuint16 cur_op = 0;
+                if (has_return_value) {
+                    if (box_return_value) {
+                        rb_ins->operands[cur_op] = MVM_spesh_manipulate_get_temp_reg(tc, g, MVM_reg_int64);
+                    }
+                    else {
+                        rb_ins->operands[cur_op] = ins->operands[0];
+                    }
+                    MVM_spesh_get_facts(tc, g, rb_ins->operands[cur_op])->writer = rb_ins;
+                    cur_op++;
+                }
+
+                /* Add the operand with the bytecode to run. */
+                rb_ins->operands[cur_op] = temporaries[op->res_code.temp_invokee];
+                MVM_spesh_usages_add_by_reg(tc, g, rb_ins->operands[cur_op], rb_ins);
+                cur_op++;
+
+                MVMuint8 *is_rw_operand = MVM_spesh_alloc(tc, g, rb_op->num_operands * sizeof(MVMuint8));
+                memset(is_rw_operand, 0, rb_op->num_operands * sizeof(MVMuint8));
+                MVMSpeshOperand *rw_operands = MVM_spesh_alloc(tc, g, rb_op->num_operands * sizeof(MVMSpeshOperand));
+
+                MVMSpeshFacts *object_facts = MVM_spesh_get_facts(tc, g, temporaries[op->res_code.temp_invokee]);
+                MVMNativeCallBody *body = MVM_nativecall_get_nc_body(tc, object_facts->value.o);
+                if (object_facts->flags & MVM_SPESH_FACT_KNOWN_VALUE) {
```

Comment:
Not sure it's safe to use `object_facts->value.o` before checking it's a known value.